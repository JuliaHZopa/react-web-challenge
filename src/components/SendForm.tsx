import * as React from "react";
import { connect } from "react-redux";
import {Dispatch, bindActionCreators} from "redux";

import { makeTransaction, updateTransactionData } from "../actions/transactionActions";
import { updateErrors } from "../actions/updateErrorsActions";
import { validateField } from "../validation/SendFormValidation";
import { saveTransaction, setCurrentAmount } from "../actions/accountActions";
import Field from "./Field";

interface ISendFormProps {
  actions: any;
}

interface IStateProps {
  transaction: {
    amount: number;
    reciever_name: string;
    reciever_email: string;
  }
  account: {
    currentAmount: number;
  };
  errors: any;
}

const validationMapping = {
  amount: ['isRequired', 'isLargerThanZero'],
  reciever_email: ['isRequired', 'isValidEmail'],
  reciever_name: ['isRequired', 'minLength2'],
}

class SendForm extends React.Component<ISendFormProps & IStateProps, any> {

  public constructor(props: ISendFormProps & IStateProps) {
    super(props);
  }

  public render() {
    return (
      <form>
        <Field
          type='text'
          name='reciever_name'
          value={this.props.transaction.reciever_name}
          handleOnChange={this.handleOnChange}
          placeholder={`Recipient's name`}
          label={`Recipient's name`}/>
        <Field
          type='text'
          name='reciever_email'
          value={this.props.transaction.reciever_email}
          handleOnChange={this.handleOnChange}
          placeholder={`Recipient's email`}
          label={`Recipient's email`}/>
        <Field
          type='number'
          name='amount'
          value={this.props.transaction.amount}
          handleOnChange={this.handleOnChange}
          placeholder={`Amount to send`}
          label={`Amount to send`}/>
        <button onClick={this.handleOnClick}>Send money</button>
      </form>
    )
  }

  private handleOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const fieldName = e.target.name;
    const data = {
      key: fieldName,
      value: e.target.value
    }

    this.props.actions.updateTransactionData(data)

    const validationResult = validateField(fieldName, e.target.value, validationMapping[fieldName]);
    this.props.actions.updateErrors(validationResult);
  }

  private isFormValid = () => {
    if (!this.props) {
      return false;
    }
    const errors = this.props.errors ? this.props.errors : false;
    return errors ? Object.keys(errors).every((error: string) => this.props.errors[error] === true) : false;
  }

  private handleOnClick = (e: React.MouseEvent) => {
    const currentAmount = this.props.account.currentAmount;
    e.preventDefault();
    if (this.isFormValid() && this.props.transaction.amount <= this.props.account.currentAmount) {

      // TODO: handleOnClick should not handle all of this.
      // TODO: Not here but add support for amountSent
      this.props.actions.makeTransaction(this.props.transaction);
      this.props.actions.saveTransaction(this.props.transaction);

      const newAmout = currentAmount - this.props.transaction.amount!;
      this.props.actions.setCurrentAmount(newAmout);

      this.props.actions.updateTransactionData({
        transaction: {
          amount: undefined,
          reciever_name: undefined,
          reciever_email: undefined,
        },
        errors: undefined
      });
    };
  }
}

const mapStateToProps = (state: IStateProps) => {
  return {
    transaction: {
      amount: state.transaction.amount,
      reciever_name: state.transaction.reciever_name,
      reciever_email: state.transaction.reciever_email,
    },
    account: {
      currentAmount: state.account.currentAmount
    },
    errors: state.errors,
  }
}

const mapDispatchToProps = (dispatch: Dispatch) => {
  return {
    actions: bindActionCreators({
      updateTransactionData,
      updateErrors,
      makeTransaction,
      saveTransaction,
      setCurrentAmount
    }, dispatch)
  }
};
â€‹
export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SendForm)
