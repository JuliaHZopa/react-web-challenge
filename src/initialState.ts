export interface IInitialState {
  transaction: ITransaction;
  account: IAccount;
  errors?: object;
}

export interface ITransaction {
  amount?: number;
  reciever_name?: string;
  reciever_email?: string;
}

export interface IAccount {
  currentAmount: number,
  amountSent?: number;
  previousTransactions?: IPreviousTransactions[];
}

export interface IPreviousTransactions {
  amount: number;
  reciever_name: string;
  reciever_email: string;
}

const initialState = {
  account: {
    currentAmount: undefined,
    amountSent: undefined,
  },
  transaction: {
    amount: undefined,
    reciever_name: undefined,
    reciever_email: undefined,
  },
  errors: undefined,
}

export default initialState;
