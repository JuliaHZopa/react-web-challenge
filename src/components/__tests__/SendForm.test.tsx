import * as React from "react";
import * as renderer from "react-test-renderer";
import { Provider } from "react-redux";
import initialState from "../../initialState";
import SendForm from "../SendForm";
import configureStore from 'redux-mock-store';

const middlewares: any = []
const mockStore = configureStore(middlewares)

const store = mockStore(initialState);

describe('SendForm', () => {
  test('it renders', () => {
    const render = renderer.create(<Provider store={store}><SendForm /></Provider>).toJSON();
    expect(render).toMatchSnapshot();
  });

  test('it calls handleOnChange on form change', () => {
    // TODO: write test
  });

  test('it calls updateTransactionData on form change', () => {
    // TODO: write test
  });

  test('it calls ValidateField with field name and value', () => {
    // TODO: write test
  });

  test('ValidateField returns an object containing a key and a value', () => {
    // TODO: write test
  });

  test('it calls updateErrors with an object containing a key and a value', () => {
    // TODO: write test
  });

  test('it calls handleOnClick when the submit button is clicked', () => {
    // TODO: write test
  });

  test('it calls isFormValid when handleOnClick is called', () => {
    // TODO: write test
  });

  test('it calls makeTransaction if handleOnClick is called and the form is valid', () => {
    // TODO: write test
  });

  test('it calls resets the form if handleOnClick is called and the form is valid', () => {
    // TODO: write test
  });

  test('it returns false if isFormValid is called without a state', () => {
    // TODO: write test
  });

  test('it returns true if isFormValid is called and the form is valid', () => {
    // TODO: write test
  });

  test(`it returns false if isFormValid is called and this.state.erros is undefined`, () => {
    // TODO: write test
  });

  test('it returns false isFormValid is called and form is invalid', () => {
    // TODO: write test
  });
})
