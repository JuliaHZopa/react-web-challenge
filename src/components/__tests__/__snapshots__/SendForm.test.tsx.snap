// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendForm it renders 1`] = `
<form>
  <div
    className="field"
  >
    <label>
      Recipient's name
    </label>
    <input
      name="recipient_name"
      onChange={[Function]}
      placeholder="Recipient's name"
      type="text"
    />
  </div>
  <div
    className="field"
  >
    <label>
      Recipient's email
    </label>
    <input
      name="reciever_email"
      onChange={[Function]}
      placeholder="Recipient's email"
      type="text"
    />
  </div>
  <div
    className="field"
  >
    <label>
      Amount to send
    </label>
    <input
      name="amount_to_send"
      onChange={[Function]}
      placeholder="Amount to send"
      type="number"
    />
  </div>
  <button
    onClick={[Function]}
  >
    Send money
  </button>
</form>
`;
